Allow inducing G-sets along group homomorphisms

lgoettgens opened the pull request:
As discussed earlier today with @ThomasBreuer. I tried to align the name and argument order with the methods for `GAPClassFunction` and `GModule`.
Due to not all group types having usable morphism types, I added an undocumented method for general maps between groups, which should get adapted once we have proper group homes.

@jamesnohilly this allows for a similar cleanup of `action_homomorphism` in #4609 as for`FinGenAbGroup`s in this PR. Depending on the order these two PRs get merged, the later one should perform the cleanup.

lgoettgens committed:
Allow inducing G-sets along group homomorphisms

lgoettgens committed:
Extend to groups with weird hom types

ThomasBreuer commented:
In the example
```
G = symmetric_group(4)
Omega = gset(G, permuted, [[1,1,2,3]])
H = permutation_group(8, [cperm([1,3], [2,4]), cperm([1,5], [2,6], [3,7], [4,8])])
phi = hom(H, G, [cperm([1,2]), cperm([1,3], [2,4])])
Omega2 = induce(Omega, phi)
```
my interpretation of the documentation is that `Omega` and `Omega2` shall be equal as sets, they just carry different actions. `G` acts on `Omega`, `H` acts on `Omega2`.
For that, we cannot take just the seeds of `Omega` (the points whose closure under the action of `G` is `Omega`) as the seeds of `Omega2`, because `H` can be (and in the example is) smaller.
(If `phi` is known to be surjective then taking the seeds of `Omega` as the seeds of `Omega2` is sufficient.
The check whether `acting_group(Omega)` equals `codomain(phi)` does not guarantee surjectivity.)

We could "unpack" `Omega` by computing all its elements, and take them as the seeds for the intended `H`-set.
If that is not desirable, we could think of creating the orbits of `phi(H)` on `Omega`, and then taking orbit representatives as seeds for `Omega2`. (Then we can also set the flag that `Omega2` is closed under the action.)

(If we are instead interested in an `H`-orbit of a particular element `omega` of `Omega`, w.r.t. the action induced by `phi`, then this is something different. For that, we can first create the action function for `H` from `phi` and the action function of `Omega`, and then create the `H`-orbit of `omega` w.r.t. that function. In the above example, `omega` is `[1,1,2,3]`, and the `H`-orbit of `omega` is `Omega2`.)

lgoettgens committed:
Keep underlying set the same

lgoettgens commented:
> my interpretation of the documentation is that `Omega` and `Omega2` shall be equal as sets, they just carry different actions. `G` acts on `Omega`, `H` acts on `Omega2`.
> For that, we cannot take just the seeds of `Omega` (the points whose closure under the action of `G` is `Omega`) as the seeds of `Omega2`, because `H` can be (and in the example is) smaller.

Indeed, good catch. I opted to use `Omega` as an iterable of the seeds of `Omega2` (with `closed=true`) as was already sone previously in the now replaced `action_homomorphism(GSetByElements{FinGenAbGroup}))`.



> (If we are instead interested in an `H`-orbit of a particular element `omega` of `Omega`, w.r.t. the action induced by `phi`, then this is something different. For that, we can first create the action function for `H` from `phi` and the action function of `Omega`, and then create the `H`-orbit of `omega` w.r.t. that function. In the above example, `omega` is `[1,1,2,3]`, and the `H`-orbit of `omega` is `Omega2`.)

I am not quite sure which of the two scenarios is the correct one for my use-case (in the cases of non-surjective homs). But just to make sure: With the now updated code, I could compute the `H`-orbit of some `omega` by `orbit(Omega2, omega)`, right?

ThomasBreuer commented:
Concerning the idea to document `induce` only for the case that the map `phi` is a `GAPGroupHomomorphism{U, T}` but to allow also for a `Map{U, T}`, I thought that a `Map` between groups is always assumed to be structure preserving (i. e., a group homomorphism).

The manual is not good enough in this respect: The section "Group homomorphisms" talks only about `GAPGroupHomomorphism{S, T}`, and the section "Maps" belongs to "Groups"/"Abelian Groups"/"Morphisms".

ThomasBreuer commented:
> I am not quite sure which of the two scenarios is the correct one for my use-case (in the cases of non-surjective homs). But just to make sure: With the now updated code, I could compute the H-orbit of some omega by orbit(Omega2, omega), right?

Yes.

However, if one is interested just in one particular H-orbit, it might be cheaper to compute this directly, without creating the perhaps much too large H-set `Omega2`. For example, if H is the trivial group then computing the H-orbit of an element in `Omega` is cheap, but describing `Omega` as an H-set can be expensive. (We need a better way to describe `Omega` as an H-set, relative to the known G-set, and without the need to write down all elements.)

ThomasBreuer reviewed (approved):
Looks good. Thanks.

lgoettgens commented:
> However, if one is interested just in one particular H-orbit, it might be cheaper to compute this directly, without creating the perhaps much too large H-set `Omega2`. For example, if H is the trivial group then computing the H-orbit of an element in `Omega` is cheap, but describing `Omega` as an H-set can be expensive. (We need a better way to describe `Omega` as an H-set, relative to the known G-set, and without the need to write down all elements.)

Indeed. My idea for that would be to add a function `induced_action_function(Omega, phi)` that returns the `fun` object from this PR, and then rewrite `_induce` to make use of that action function.
If we export that function as well, one could make use of it as `orbit(H, fun, omega)` for the `H`-orbit of `omega`.
What dou you think about this idea @ThomasBreuer ?

ThomasBreuer commented:
> My idea for that would be to add a function induced_action_function(Omega, phi) that returns the fun object from this PR, and then rewrite _induce to make use of that action function.
If we export that function as well, one could make use of it as orbit(H, fun, omega) for the H-orbit of omega.
What do you think about this idea @ThomasBreuer ?

Yes.
We discussed yesterday that for some purposes, it is good to work with explicit G-sets, and for other purposes, it is good to work directly with their ingredients (the points, the group, the function). Thus it makes sense to switch between the two viewpoints.

lgoettgens committed:
Add `induced_action` and `induced_action_function`

lgoettgens commented:
I now added `induced_action` and `induced_action_function` where the latter is just a shorthand for `induced_action(action_function(Omega), phi)`. It is just there for convenience.

ThomasBreuer reviewed (approved):
Good.
(I am a bit surprised that comparing the two functions with `==` works.)

lgoettgens committed:
Add a comment
